ZIP 400: Update wallet schema (#968)
zcash/zips: zips/zip-0400.rst

Reviewer: Lines 12-17: MUST and MAY are not used (and don't appear to be needed). Just delete the whole Terminology section if there are no other terms to define.

Schema
------

The following key-values can be found: the property names in **bold** mean only one instance of this type can 
exist in the entire database, while the others, suffixed by '\*' can have multiple instances. Keys and Values columns of the table

> contain the types that the stored data is representing. The Version column indicates the zcashd version the key is present. 

Reviewer: ```suggestion
contain the types that the stored data is representing. The Version column indicates the zcashd version(s) that create databases containing this key.
```
The zcashd versions that create a key could potentially be different from the versions that understand it.

Included also there are the variable names hoping it will add some clarity to what the stored data is representing.

.. csv-table::
   :header: "Version", "Name", "Description", "Keys", "Values"
   :align: left

>    "3..4", "acc*", "Account data", "1. ``string strAccount``", "1. ``CAccount account``"

Reviewer: If the end version is less than 6:
* Check this is actually correct. 
* If it is, specify how later version handle/migrate the key.

Author: Looks like both `acc` and `acentry`, which were part of the 'account' API, were disabled since the Zcash launch, and removed in v4.5.0: https://github.com/zcash/zcash/releases/tag/v4.5.0.

Here is also the Bitcoin Core release it references: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.17.0.md#label-and-account-apis-for-wallet.

Should we add the fact that both were deprecated/disabled, and that they have no replacement?

>    "3..4", "acc*", "Account data. This entry is used in Bitcoin and not in Zcash.", "1. ``string strAccount``", "1. ``CAccount account``"
>    "3..4", "acentry*", "Track internal transfers between accounts in the same wallet. This entry is used in Bitcoin and not in Zcash.", "1. ``string strAccount``
>    2. ``uint64_t nNumber``", "1. ``CAccountingEntry acentry``"

Reviewer: Does this mean that these fields will never be present in a Zcash walletdb? In that case just delete them.
```suggestion
```

   "3..4", "**chdseed**", "Encrypted HD seed. Replaced by ``cmnemonicphrase`` in later versions.", "1. ``uint256 seedFp``", "1. ``vector<unsigned char> vchCryptedSecret``"
   "3..6", "ckey*", "Encrypted transparent pubkey and private key.", "1. ``vector<unsigned char> vchPubKey``", "1. ``vector<unsigned char> vchPrivKey``"
   "3..6", "csapzkey*", "Encrypted Sapling pubkey and private key.", "1. ``libzcash::SaplingIncomingViewingKey ivk``", "1. ``libzcash::SaplingExtendedFullViewingKey extfvk`` 
   2. ``vector<unsigned char> vchCryptedSecret``"
   "3..6", "cscript\*", "Serialized redeem script, used inside transaction inputs and outputs", "1. ``uint160 hash``", "1. ``CScript script``"
   "3..6", "czkey*", "Encrypted Sprout pubkey and private key.", "1. ``libzcash::SproutPaymentAddress addr``", "1. ``uint256 rkValue`` 
   2. ``vector<unsigned char> vchCryptedSecret``"
   "3..6", "**defaultkey**", "Default Transparent key.", "", "1. ``CPubKey CWallet::vchDefaultKey``"
   "3..6", "destdata*", "Adds a destination data tuple to the store.", "1. ``std::string strAddress``
   2. ``std::string strKey``", "1. ``std::string strValue``"

>    "3..6", "**hdchain**", "Hierarchical Deterministic chain code, derived from   seed.", "", "1. ``CHDChain chain``"

Reviewer: ```suggestion
   "3..6", "**hdchain**", "Hierarchical Deterministic chain code, derived from seed.", "", "1. ``CHDChain chain``"
```

   "3..4", "hdseed*", "Hierarchical Deterministic seed. [#zip-0032]_ Replaced by ``mnemonicphrase`` in later versions.", "1. ``uint256 seedFp``", "1. ``RawHDSeed rawSeed``"
   "3..6", "key*", "Transparent pubkey and privkey.", "1. ``CPubKey vchPubKey``", "1. ``CPrivKey pkey``"
   "3..6", "keymeta*", "Transparent key metadata.", "1. ``CPubKey vchPubKey``", "1. ``CKeyMetadata keyMeta``"

>    "3..6", "**minversion**", "Wallet required minimal version.", "", ""

Reviewer: What is the type of this?

Author: Updated!

   "3..6", "**mkey**", "Master key, used to encrypt public and private keys of the database.", "1. ``unsigned int nID``", "1. ``CMasterKey kMasterKey``"
   "3..6", "name*", "Name of an address to insert in the address book.", "1. ``string strAddress``", "1. ``string strAddress``"
   "3..6", "**orderposnext**", "Index of next tx.", "", "1. ``int64_t nOrderPosNext``"
   "3..6", "pool*", "", "1. ``int64_t nIndex``", "1. ``CKeyPool keypool``"
   "3..6", "purpose*", "Short description or identifier of an address.", "1. ``string strAddress``", "1. ``string strPurpose``"
   "3..6", "sapzaddr*", "Sapling z-addr Incoming Viewing key and address.", "1. ``libzcash::SaplingPaymentAddress addr``", "1. ``libzcash::SaplingIncomingViewingKey ivk``"

>    "3..6", "sapextfvk*", "Sapling Extended Full Viewing Key", "", ""

Reviewer: What is the type of this?

Author: Updated! However, due to the length of the type, a horizontal scrollbar appeared as the table became wider.

   "3..6", "sapzkey*", "Sapling Incoming Viewing Key and Extended Spending Key", "1. ``libzcash::SaplingIncomingViewingKey ivk``", "1. ``libzcash::SaplingExtendedSpendingKey key``"

>    "3..6", "tx*", "Store all transactions that are related to wallet.", "1. ``uint256 hash``", "1. ``CWalletTx wtx``"

Reviewer: ```suggestion
   "3..6", "tx*", "All transactions that are related to the wallet.", "1. ``uint256 hash``", "1. ``CWalletTx wtx``"
```

   "3..6", "**version**", "The ``CLIENT_VERSION`` from ``clientversion.h``.", "", "1. ``int nFileVersion``"
   "3..6", "vkey*", "Sprout Viewing Keys.", "1. ``libzcash::SproutViewingKey vk``", "1. ``char fYes``"
   "3..6", "watchs*", "Watch-only t-addresses.", "1. ``CScript script``", "1. ``char fYes``"
   "3..6", "**witnesscachesize**", "Shielded Note Witness cache size.", "", "1. ``int64_t nWitnessCacheSize``"

>    "3..6", "wkey*", "Wallet key.", "", ""

Reviewer: What is the type of this?

> .. [#ZIP400Issue] `ZIP 400 issue <https://github.com/zcash/zips/issues/350>`_

Reviewer: ```suggestion
.. [#zips-issue350] `zcash/zips issue #350: [ZIP 400] Format of Zcash wallet.dat files <https://github.com/zcash/zips/issues/350>`_
```

.. [#zip-0032] `ZIP 32: Shielded Hierarchical Deterministic Wallets <zip-0032.rst>`_
.. [#CrypterCode] `Database key encryption implementation <https://github.com/zcash/zcash/blob/master/src/wallet/crypter.h>`_

> .. [#UFVKEncoding] `Zcash Protocol Specification, Version 2023.4.0. Section 5.6.4: Unified and Orchard Encodings <https://zips.z.cash/protocol/protocol.pdf#orchardencodings>`_

Reviewer: ```suggestion
.. [#UFVKEncoding] `Zcash Protocol Specification, Version 2024.5.1. Section 5.6.4: Unified and Orchard Encodings <https://zips.z.cash/protocol/protocol.pdf#orchardencodings>`_
```
